### PR TITLE
Use default AWS credential handling under normal circumstances

### LINKED ISSUE #191: AWS credentials not handled correctly
There are various issues with the AWS credential handling when downloading from S3. #185 #157 #152 #22 https://github.com/gruntwork-io/terragrunt/pull/775

I'm less familiar with the AWS SDK for Go, but in my experience with those for JavaScript and Python, the credential handling here [1] looks like it can be made simpler and more robust by using the SDK as documented [2].

The particulars about communicating with the EC2 (or ECS) metadata API, reading the shared config or credentials file, choosing a region, etc., are all handled by the SDK's default credential provider chain. The explicit provider chain constructed in `getAWSConfig` lacks support for various edge cases.

Would you be open to a PR changing the way AWS sessions are created?

[1] https://github.com/hashicorp/go-getter/blob/da0323b9545e4b9d821e9c7d8ed4429e41d997ed/get_s3.go#L182-L217
[2] "Configuring Credentials" section on https://docs.aws.amazon.com/sdk-for-go/api/

### PR DESCRIPTION
**Pull request by jbergknoff-rival**
Fixes #191 (and #185, #157, #152, #22)

This PR changes go-getter to default to the AWS SDK's built in credential resolution. This makes it possible for go-getter to download from S3 using an assumed role (e.g. when setting environment variables `AWS_SDK_LOAD_CONFIG=true AWS_PROFILE=...`, go-getter will assume that role before attempting any operations). It should also add support for various other edge cases relating to AWS credentials.

The one case where go-getter will continue to construct its own credential chain is the one with an override for the instance metadata URL. I'm not familiar with the use case of overriding the instance metadata URL, so I left that code as it was.

I'd like to add a test to protect against regressions in assume-role behavior, but I'd appreciate some guidance there. I see the hardcoded read-only IAM access key/secret. Do those credentials also have the ability to assume some role?

### CODE REVIEW

File: get_s3.go
Comment (jbergknoff-rival) [inline get_s3.go]: This is just a useful configuration change to help debugging credentials issues.

**Before:**
```
Copying configuration from "s3::https://s3-us-west-2.amazonaws.com/..."...

Error: Failed to download module

Could not download module "root"
(-from-module="s3::https://s3-us-west-2.amazonaws.com/...":1)
source code from
"s3::https://s3-us-west-2.amazonaws.com/...
NoCredentialProviders: no valid providers in chain. Deprecated.
    For verbose messaging see aws.Config.CredentialsChainVerboseErrors
```

**After:**
```
Copying configuration from "s3::https://s3-us-west-2.amazonaws.com/..."...

Error: Failed to download module

Could not download module "root"
(-from-module="s3::https://s3-us-west-2.amazonaws.com/...":1)
source code from
"s3::https://s3-us-west-2.amazonaws.com/...":
NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: failed to find credentials in the
environment.
SharedCredsLoad: failed to load profile, <profile name>.
EC2RoleRequestError: no EC2 instance role found
caused by: RequestError: send request failed
caused by: Get
http://169.254.169.254/latest/meta-data/iam/security-credentials/: dial tcp
169.254.169.254:80: connect: no route to host
```

Comment (bashims) [reply get_s3.go]: This looks good to me and obsoletes my patch. We will test it under ECS to make sure. Thanks!

**Comment by NinaDenisova** [comment]
Hi @jbergknoff-rival I also ran into the issue of not being able to access s3 buckets when it belongs to an originalizational unit. I tested your PR and it does solve the issue for us as well. If I understand it correctly, the next step would be to sign the SLA so that the PR can be merged. I was wondering if you are planning to sign it in the near future. 
Thank you!

**Comment by jbergknoff-rival** [comment]
Oh, thanks for pointing that out, @NinaDenisova . I actually have signed the CLA, and when I visit https://cla.hashicorp.com/hashicorp/go-getter?pullRequest=218, it shows me

image

and the form at the bottom is grayed out, with my details filled in. Maybe there's a bug on the status check.

Either way, I don't think that was the primary blocker on this PR. It just hasn't received any attention from maintainers.

**Comment by NinaDenisova** [comment]
Thank you @jbergknoff-rival for the quick reply! 
@hashicorp-cla Could you please clarify the status of CLA for this PR? 
@hashicorp-cloud I was wondering if anybody has already checked this PR as it's been open for a while now. The changes are not extensive and enable letting AWS take care of credentials provider chain rather than creating the chain manually for the case when credentials are not supplied. This addresses a few issues stated in the PR description (Fixes #191 (and #185, #157, #152, #22)). 
Thank you, everybody.

**Comment by azr** [comment]
Hello there, about the CLA; you need to make sure the email you signed the CLA with is the same as the email you have in your commit, otherwise we cannot merge this.

**Comment by jbergknoff-rival** [comment]
Thanks for the clarification. Now I understand what's going on. At the time I signed the CLA, the email address of the commit was associated with this GitHub account. The company has since become defunct, and I disassociated this GitHub account from that email address (which is also defunct). If I get a chance, I'll try rewriting the commit history on the PR branch.

**Review by azr — APPROVED** [review]
LGTM ( just waiting for CLA )

**Comment by jbergknoff** [comment]
:+1: I rewrote the commit to use the appropriate email address and now the CLA check is there. I also rebased from the current master branch.

**Comment by azr** [comment]
Awesome !

**Comment by apeeters** [comment]
How does this relate to / duplicates  https://github.com/hashicorp/go-getter/pull/261?

**Pull request merged**
